### PR TITLE
KBV-300: Request the auth token via the private_key_jwt method

### PR DESCRIPTION
**Pull request by idrop**
### What changed
Refactored code to fetch the auth token via the **private_key_jwt** method

### Why did it change
As per the [RFC](https://github.com/alphagov/digital-identity-architecture/blob/main/rfc/0001-crossteam-oauth-profile.md#requests-to-the-token-endpoint), it has been decided that the clients must authenticate to the authorization server's token endpoint using [the private_key_jwt method](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication)

### Issue tracking

- [KBV-300](https://govukverify.atlassian.net/browse/KBV-300)
- [RFC](https://github.com/alphagov/digital-identity-architecture/blob/main/rfc/0001-crossteam-oauth-profile.md#client-profile)

### CODE REVIEW

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java [hunk 1]
```diff
@@ -27,18 +41,30 @@
...
-
-    public AccessTokenService(DataStore<AddressSessionItem> dataStore, Long bearerAccessTokenTtl) {
+    private final ConfigurationService configurationService;
+    private final JWTVerifier jwtVerifier;
+
+    public AccessTokenService(
+            DataStore<AddressSessionItem> dataStore,
+            Long bearerAccessTokenTtl,
+            ConfigurationService configurationService,
+            JWTVerifier jwtVerifier) {
```
Comment (idrop) [inline lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java:55]: added JWTVerifier as its now used in the AddressSessionService and AddressTokenService

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java [hunk 2]
```diff
@@ -88,41 +90,115 @@ public void writeToken(
...
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(CODE, Collections.emptyList()));
-        var redirectUri =
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(REDIRECT_URI, Collections.emptyList()));
-        var grantType =
-                listHelper.getValueOrThrow(
-                        queryParameters.getOrDefault(GRANT_TYPE, Collections.emptyList()));
+            ClientAuthentication clientAuthentication = tokenRequest.getClientAuthentication();
+            PrivateKeyJWT privateKeyJWT = (PrivateKeyJWT) clientAuthentication;
```
Comment (idrop) [inline lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java:119]: @amanm-alphagov this is the important bit of the PR - where we get a PrivateKeyJWT, then verify its payload and signing

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.security.cert.CertificateFactory;
+import java.security.interfaces.ECPublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.HashSet;
+import java.util.Map;
+
+public class JWTVerifier {
```
Comment (idrop) [inline lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java:27]: @amanm-alphagov sorry but we'll need tests for this new class - you can lift the tests from AddressSessionServiceTest that build up JWTs

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java (hunk 1 above)
Comment (amanm-alphagov) [reply lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java:52]: added unit tests for JWTVerifier.class

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java (hunk 3 above)
Comment (amanm-alphagov) [reply lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/JWTVerifier.java:27]: added unit tests for JWTVerifier class

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/address/api/handler/AccessTokenHandlerTest.java [hunk 4]
```diff
@@ -55,18 +55,25 @@ void setUp() {
...
+        String authCodeValue = "12345";
+        String grantType = "authorization_code";
         String tokenRequestBody =
-                "code=12345&redirect_uri=http://test.com&grant_type=authorization_code&client_id=test_client_id";
+                String.format(
+                        "code=%s"
+                                + "&client_assertion=%s"
+                                + "&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
+                                + "&client_id=urn:uuid:ipv-core"
+                                + "&grant_type=%s",
```
Comment (jkunle) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/address/api/handler/AccessTokenHandlerTest.java:66]: This is really weird as you are using the `String.format` you shouldn't have to use `+`, I think you can replace the `+` with a commas for the same effect

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java [hunk 5]
```diff
@@ -1,83 +1,88 @@
...
-
-    public AccessTokenService(DataStore<AddressSessionItem> dataStore, Long bearerAccessTokenTtl) {
+    private final ConfigurationService configurationService;
+    private final JWTVerifier jwtVerifier;
+    private ListUtil listHelper = new ListUtil();
+
+    public AccessTokenService(
+            DataStore<AddressSessionItem> dataStore,
+            Long bearerAccessTokenTtl,
+            ConfigurationService configurationService,
```
Comment (jkunle) [inline lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java:51]: I think `bearerAccessTokenTtl` comes from `configurationService.getBearerAccessTokenTtl` at some stage this class only needed  `bearerAccessTokenTtl` hence there was no reason to pass in the entire configurationService. I think we can remove  `bearerAccessTokenTtl`

**Comment by jkunle** [comment]
Please have look at the bug and code smell on offered by SonarCloud

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java [hunk 6]
```diff
@@ -88,47 +93,119 @@ public void writeToken(
...
+                    "Authorisation code does not match with authorization Code for Address Session Item");
         }
-        if (!URI.create(redirectUri).equals(addressSessionItem.getRedirectUri())) {
-            throw new AccessTokenRequestException(
-                    String.format(
-                            "Requested redirectUri: %s does not match existing redirectUri: %s",
-                            redirectUri, addressSessionItem),
-                    OAuth2Error.INVALID_GRANT);
+
+        Issuer jwtIssuer = privateKeyJWT.getJWTAuthenticationClaimsSet().getIssuer();
```
[the remark this answers is not in the document]
Comment (techaman-m) [reply lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java:205]: Added a new method with more micro grained level checks.

File: lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java (hunk 5 above)
Comment (techaman-m) [reply lib/src/main/java/uk/gov/di/ipv/cri/address/library/service/AccessTokenService.java]: Removed.

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/address/api/handler/AccessTokenHandlerTest.java (hunk 4 above)
Comment (techaman-m) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/address/api/handler/AccessTokenHandlerTest.java:66]: I think you can use either `+` or `,` depending on the need. The `+` is just a string concatenation operator in this case and it is there to split the string. We can put all the strings one line. Using `,` the same thing can be achieved but the way the params need to be passed will have to change.

**Review by jkunle — APPROVED** [review]

**Pull request merged**